src/profile-logic/import/instruments/index.js

```diff
@@ -0,0 +1,926 @@
+    const schemaFile = storedir.files.get('schema.xml');
+    if (!schemaFile) {
+      continue;
+    }
+    const schema = await readAsText(schemaFile);
+    if (!/name="time-profile"/.exec(schema)) {
+      continue;
+    }
+    const bulkstore = new BinReader(
+      await readAsArrayBuffer(ensureExists(storedir.files.get('bulkstore')))
+    );
+    // Ignore the first 3 words
+    bulkstore.readUint32();
+    bulkstore.readUint32();
+    bulkstore.readUint32();
+    const headerSize = bulkstore.readUint32();
+    const bytesPerEntry = bulkstore.readUint32();
+
+    bulkstore.seek(headerSize);
+
+    const samples: InstrumentsSample[] = [];
+    while (true) {
+      // Schema as of Instruments 8.3.3 is a 6 byte timestamp, followed by a bunch
+      // of stuff we don't care about, followed by a 4 byte backtrace ID
+      const timestamp = bulkstore.readUint48();
+      if (timestamp === 0) {
+        break;
+      }
+
+      const threadID = bulkstore.readUint32();
+
+      bulkstore.skip(bytesPerEntry - 6 - 4 - 4);
+      const backtraceID = bulkstore.readUint32();
+      samples.push({ timestamp, threadID, backtraceID });
+    }
+    return samples;
+  }
+  throw new Error('Could not find sample list');
+}
+
+// This function extracts core directory file which contains information about samples
+// path: tree.subdirectories.corespace.subdirectories.run${runNumber}.core
+function getCoreDirForRun(
+  tree: TraceDirectoryTree,
+  selectedRun: number
+): TraceDirectoryTree {
+  const corespace = ensureExists(tree.subdirectories.get('corespace'));
+  const corespaceRunDir = ensureExists(
+    corespace.subdirectories.get(`run${selectedRun}`)
+  );
+  return ensureExists(corespaceRunDir.subdirectories.get('core'));
+}
+
+// This function mainly extracts two entities: samples and arrays
+// Here arrays contains all the information about stack trace at a given timestamp.
+// Each samples has a field named 'backtraceID' which is an index into arrays
+// Iterating recursively into arrays by given backtraceID it extracts a backtraceStack for each sample
+async function getSamples(args: {
+  tree: TraceDirectoryTree,
+  addressToFrameMap: Map<number, InstrumentsFrameInfo>,
+  runNumber: number,
+}): Promise<{
+  samples: Array<InstrumentsSample>,
+  backtraceIDtoStack: Map<number, Array<number>>,
+}> {
+  const { tree, addressToFrameMap, runNumber } = args;
+  const core = getCoreDirForRun(tree, runNumber);
+  const samples = await getRawSampleList(core);
+  const arrays = await getIntegerArrays(core);
+  const backtraceIDtoStack = new Map<number, Array<number>>();
+
+  function appendRecursive(k: number, stack: Array<number>) {
+    const frame = addressToFrameMap.get(k);
+    if (frame) {
+      stack.push(k);
+      return;
+    }
+
+    const arrayIndex = k & 0xffffffff;
+    if (arrayIndex in arrays) {
+      for (const addr of arrays[arrayIndex]) {
+        appendRecursive(addr, stack);
+      }
+      return;
+    }
```

Comment:
Could we get some better names here, and maybe a comment on how this structure works?
So the result is a "backtraceIDToStack" map. In the lambda below, the keys in this map are called "id". And "id" is passed to the "k" argument of appendRecursive. And then k is used as a key to do a lookup in addressToFrameMap. So is k a backtraceID or an address?

And then appendRecursive is both recursive and has a for-of loop in it. It seems like there's a tree structure underlying here, where each backtraceID refers to a list of other backtraceIDs (let's call them children), and appendRecursive traverses the subtree of a given backtraceID in a depth-first search. So the "stack" that's output is really a list of the nodes that were encountered, in traversal order, during this DFS. This is quite unexpected! Is there some intuitive explanation of this graph structure?